c7n/actions/network.py

```diff
@@ -137,6 +161,30 @@ def get_groups_by_names(self, names):
                 names=list(unresolved), groups=[g['GroupId'] for g in sgs]))
         return sgs
 
+    def get_groups_by_tag(self, key, values, vpc_id):
+        """Get security groups that match tag values."""
+        client = utils.local_session(
+            self.manager.session_factory).client('ec2')
+        sgs = self.manager.retry(
+            client.describe_security_groups,
+            Filters=[{
+                'Name': 'tag:' + key, 'Values': values},{
+                'Name': 'vpc-id', 'Values': [vpc_id]}]).get(
+                    'SecurityGroups', [])
+        sg_ids = []
+        for s in sgs:
+            if s['GroupId'] not in sg_ids:
+                sg_ids.append(s['GroupId'])
+        if len(sg_ids) > 5:
+            raise PolicyExecutionError(self._format_error(
+                "policy:{policy} - the number of security groups with "
+                "the matching key-value pair exceeds 5. "
+                "key:{key}, values:{values}, found: {sg_ids}",
+                key=key,
+                values=values,
+                sg_ids=sg_ids))
```

Comment:
Hmm, I'm wondering if this check should be in `get_groups()`, either in addition to or instead of being here. Because the 5 sg limit should be most relevant when we're trying to actually update the resource, after calculating remove/add/add-by-tag.

Reply:
@ajkerrigan That makes total sense. At the end of the day, all we care about is the number of security groups to associate after calculating remove/add/add-by-tag. I moved the check to `get_groups()`.